draft-yasskin-dispatch-web-packaging.md

```diff
@@ -0,0 +1,650 @@
+   ":authority", ":path", and either `response-headers` has no "vary"
+   header (Section 7.1.4 of {{!RFC7231}}) or these header names aren't
+   listed in it.
+
+Let *origin* be the Web Origin {{!RFC6454}} of *header-list*'s ":scheme" and
+":authority" headers.
+
+{:anchor="hashing-resources"}
+Let *resource-bytes* be the result of encoding the array of
+\[*header-list*, `response-headers`, `body`] as Canonical CBOR in the
+following CDDL schema: [^ednote-figure-in-list]{:source="jyasskin"}
+
+[^ednote-figure-in-list]: This step would be inside the manifest-only block, but
+    then the code block is rendered out-of-order.
+
+~~~~~ cddl
+resource-bytes = [
+  request: [
+    *(header-name: bstr, header-value: bstr)
+  ],
+  response-headers: [
+    *(header-name: bstr, header-value: bstr)
+  ],
+  response-body: bstr
+]
+~~~~~
+
+Note that this uses the decoded header fields, not the bytes originally included
+in the package.
+
+The hashed data differs from {{SRI}}, which only hashes the body. Including the
+headers will usually prevent a package from relying on some of its contents
+being transferred as normal network responses, unless its author can guarantee
+the network won't change or reorder the headers.
+
+If the *package* contains a *manifest*:
+
+1. **TODO:** Let *origin-manifest* be the signed manifest for *origin*, found
+   by searching through *manifest*'s subpackages for a matching origin.
+
+1. Let *alg* be one of the `hash-algorithm`s within *origin-manifest*. The
+   parser SHOULD select the most collision-resistant hash algorithm. If the
+   parser also implements {{SRI}}, it SHOULD use the same order as its
+   `getPrioritizedHashFunction()` implementation.
+
+1. If the digest of *resource-bytes* using *alg* does not appear in the
+   *origin-manifest*'s `resource-hashes`\[*alg*] array, the parser MUST fail.
+
+
+Return the (decoded `response-headers`, `body`) pair.
+
+
+# Guidelines for package authors {#authoring-guidelines}
+
+Packages SHOULD consist of a single Canonical CBOR item matching the
+`webpackage` CDDL rule in {{top-level}}.
+
+Every resource's hash SHOULD appear in every array within
+`resource-hashes`: otherwise the set of valid resources will depend on
+the parser's choice of preferred hash algorithm.
+
+Security Considerations  {#security}
+=======================
+
+Signature validation is difficult.
+
+Packages with a valid signature need to be invalidated when either
+
+* the private key for any certificate in the signature's validation
+  chain is leaked, or
+* a vulnerability is discovered in the package's contents.
+
+Because packages are intended to be used offline, it's impossible to
+inject a revocation check into the critical path of using the package,
+and even in online scenarios,
+such
+[revocation checks don't actually work](https://www.imperialviolet.org/2012/02/05/crlsets.html).
+Instead, package consumers must check for a sufficiently recent set of
```

Comment:
Yes, although there'll be a cheap way to update them and to distribute those updates offline. I can make an argument that we should let a package live forever on a given device as long as that device _never_ goes online, but I'm not sure it'll convince everyone.